[NNAPI] Handle binary ops combining NHWC+NCHW in some cases

dreiss opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54701 [NNAPI] Initial flexible size support
* #54700 [NNAPI] Create get_next_operand_id
* #54699 [NNAPI] Add fixed-size assertions
* #54698 [NNAPI] Rename local variable
* #54697 [NNAPI] Use code generation to better support list input/output
* #54696 [NNAPI] Use array instead of struct for serializing ints
* #48925 Fix NNAPI for internal fbcode build
* **#48812 [NNAPI] Handle binary ops combining NHWC+NCHW in some cases**
* #48811 [NNAPI] Add support for unsqueeze, cat, and mean
* #47521 [NNAPI] Add unit test
* #54695 [NNAPI] Update support for Linear
* #47520 [NNAPI] Fix hardtanh
* #47519 [NNAPI] Fix MUL op
* #47518 [NNAPI] Remove solid weights support
* #47517 [NNAPI] Fix models with no weights
* #47407 Make bundled inputs work with quantized zero inputs

Summary:
This came up in a squeeze-and-excitation model.  Starting with an NHWC
tensor T, we perform a mean operation across H and W, giving an NxC
tensor, which (after some fully connected layers) is reshaped to
NxCx1x1, then multiplied with T.  To handle this, we detect the specific
case of a binary op with one NHWC input and one contiguous input with
H,W == 1,1 and allow the op to be applied (after transposing the
contiguous input).

Test Plan:
Unit test.

Differential Revision: [D25317939](https://our.internmc.facebook.com/intern/diff/D25317939)

dreiss committed:
[NNAPI] Handle binary ops combining NHWC+NCHW in some cases

Summary:
This came up in a squeeze-and-excitation model.  Starting with an NHWC
tensor T, we perform a mean operation across H and W, giving an NxC
tensor, which (after some fully connected layers) is reshaped to
NxCx1x1, then multiplied with T.  To handle this, we detect the specific
case of a binary op with one NHWC input and one contiguous input with
H,W == 1,1 and allow the op to be applied (after transposing the
contiguous input).

Test Plan:
Unit test.

[ghstack-poisoned]

dreiss committed:
Update on "[NNAPI] Handle binary ops combining NHWC+NCHW in some cases"


Summary:
This came up in a squeeze-and-excitation model.  Starting with an NHWC
tensor T, we perform a mean operation across H and W, giving an NxC
tensor, which (after some fully connected layers) is reshaped to
NxCx1x1, then multiplied with T.  To handle this, we detect the specific
case of a binary op with one NHWC input and one contiguous input with
H,W == 1,1 and allow the op to be applied (after transposing the
contiguous input).

Test Plan:
Unit test.

Differential Revision: [D25317939](https://our.internmc.facebook.com/intern/diff/D25317939)

[ghstack-poisoned]

dreiss committed:
Update on "[NNAPI] Handle binary ops combining NHWC+NCHW in some cases"


Summary:
This came up in a squeeze-and-excitation model.  Starting with an NHWC
tensor T, we perform a mean operation across H and W, giving an NxC
tensor, which (after some fully connected layers) is reshaped to
NxCx1x1, then multiplied with T.  To handle this, we detect the specific
case of a binary op with one NHWC input and one contiguous input with
H,W == 1,1 and allow the op to be applied (after transposing the
contiguous input).

Test Plan:
Unit test.

Differential Revision: [D25317939](https://our.internmc.facebook.com/intern/diff/D25317939)

[ghstack-poisoned]

axitkhurana commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -452,6 +464,39 @@ def get_constant_value(self, jitval, typekind=None):
+
+        out_oper = oper._replace(dim_order=DimOrder.CHANNELS_LAST)
+
+        inputs = [None] * 2
```
any reason to create a `[None, None]` list and then replacing values vs just creating `inputs = [in_id, self.add_immediate_int_vector([0, 2, 3, 1])`?

(same with outputs)

dreiss committed:
Update on "[NNAPI] Handle binary ops combining NHWC+NCHW in some cases"


Summary:
This came up in a squeeze-and-excitation model.  Starting with an NHWC
tensor T, we perform a mean operation across H and W, giving an NxC
tensor, which (after some fully connected layers) is reshaped to
NxCx1x1, then multiplied with T.  To handle this, we detect the specific
case of a binary op with one NHWC input and one contiguous input with
H,W == 1,1 and allow the op to be applied (after transposing the
contiguous input).

Test Plan:
Unit test.

Differential Revision: [D25317939](https://our.internmc.facebook.com/intern/diff/D25317939)

[ghstack-poisoned]

dreiss committed:
Update on "[NNAPI] Handle binary ops combining NHWC+NCHW in some cases"


Summary:
This came up in a squeeze-and-excitation model.  Starting with an NHWC
tensor T, we perform a mean operation across H and W, giving an NxC
tensor, which (after some fully connected layers) is reshaped to
NxCx1x1, then multiplied with T.  To handle this, we detect the specific
case of a binary op with one NHWC input and one contiguous input with
H,W == 1,1 and allow the op to be applied (after transposing the
contiguous input).

Test Plan:
Unit test.

Differential Revision: [D25317939](https://our.internmc.facebook.com/intern/diff/D25317939)

[ghstack-poisoned]

dreiss committed:
Update on "[NNAPI] Handle binary ops combining NHWC+NCHW in some cases"


Summary:
This came up in a squeeze-and-excitation model.  Starting with an NHWC
tensor T, we perform a mean operation across H and W, giving an NxC
tensor, which (after some fully connected layers) is reshaped to
NxCx1x1, then multiplied with T.  To handle this, we detect the specific
case of a binary op with one NHWC input and one contiguous input with
H,W == 1,1 and allow the op to be applied (after transposing the
contiguous input).

Test Plan:
Unit test.

Differential Revision: [D25317939](https://our.internmc.facebook.com/intern/diff/D25317939)

[ghstack-poisoned]

dreiss committed:
Update on "[NNAPI] Handle binary ops combining NHWC+NCHW in some cases"


Summary:
This came up in a squeeze-and-excitation model.  Starting with an NHWC
tensor T, we perform a mean operation across H and W, giving an NxC
tensor, which (after some fully connected layers) is reshaped to
NxCx1x1, then multiplied with T.  To handle this, we detect the specific
case of a binary op with one NHWC input and one contiguous input with
H,W == 1,1 and allow the op to be applied (after transposing the
contiguous input).

Test Plan:
Unit test.

Differential Revision: [D25317939](https://our.internmc.facebook.com/intern/diff/D25317939)

[ghstack-poisoned]

dreiss committed:
Update on "[NNAPI] Handle binary ops combining NHWC+NCHW in some cases"


Summary:
This came up in a squeeze-and-excitation model.  Starting with an NHWC
tensor T, we perform a mean operation across H and W, giving an NxC
tensor, which (after some fully connected layers) is reshaped to
NxCx1x1, then multiplied with T.  To handle this, we detect the specific
case of a binary op with one NHWC input and one contiguous input with
H,W == 1,1 and allow the op to be applied (after transposing the
contiguous input).

Test Plan:
Unit test.

Differential Revision: [D25317939](https://our.internmc.facebook.com/intern/diff/D25317939)

[ghstack-poisoned]

dreiss committed:
Update on "[NNAPI] Handle binary ops combining NHWC+NCHW in some cases"


Summary:
This came up in a squeeze-and-excitation model.  Starting with an NHWC
tensor T, we perform a mean operation across H and W, giving an NxC
tensor, which (after some fully connected layers) is reshaped to
NxCx1x1, then multiplied with T.  To handle this, we detect the specific
case of a binary op with one NHWC input and one contiguous input with
H,W == 1,1 and allow the op to be applied (after transposing the
contiguous input).

Test Plan:
Unit test.

Differential Revision: [D25317939](https://our.internmc.facebook.com/intern/diff/D25317939)

[ghstack-poisoned]

dreiss committed:
Update on "[NNAPI] Handle binary ops combining NHWC+NCHW in some cases"


Summary:
This came up in a squeeze-and-excitation model.  Starting with an NHWC
tensor T, we perform a mean operation across H and W, giving an NxC
tensor, which (after some fully connected layers) is reshaped to
NxCx1x1, then multiplied with T.  To handle this, we detect the specific
case of a binary op with one NHWC input and one contiguous input with
H,W == 1,1 and allow the op to be applied (after transposing the
contiguous input).

Test Plan:
Unit test.

Differential Revision: [D25317939](https://our.internmc.facebook.com/intern/diff/D25317939)

[ghstack-poisoned]

dreiss committed:
Update on "[NNAPI] Handle binary ops combining NHWC+NCHW in some cases"


Summary:
This came up in a squeeze-and-excitation model.  Starting with an NHWC
tensor T, we perform a mean operation across H and W, giving an NxC
tensor, which (after some fully connected layers) is reshaped to
NxCx1x1, then multiplied with T.  To handle this, we detect the specific
case of a binary op with one NHWC input and one contiguous input with
H,W == 1,1 and allow the op to be applied (after transposing the
contiguous input).

Test Plan:
Unit test.

Differential Revision: [D25317939](https://our.internmc.facebook.com/intern/diff/D25317939)

[ghstack-poisoned]

dreiss committed:
Update on "[NNAPI] Handle binary ops combining NHWC+NCHW in some cases"


Summary:
This came up in a squeeze-and-excitation model.  Starting with an NHWC
tensor T, we perform a mean operation across H and W, giving an NxC
tensor, which (after some fully connected layers) is reshaped to
NxCx1x1, then multiplied with T.  To handle this, we detect the specific
case of a binary op with one NHWC input and one contiguous input with
H,W == 1,1 and allow the op to be applied (after transposing the
contiguous input).

Test Plan:
Unit test.

Differential Revision: [D25317939](https://our.internmc.facebook.com/intern/diff/D25317939)

[ghstack-poisoned]
